lib/matplotlib/axes/_base.py

```diff
@@ -616,10 +626,21 @@ def __init__(self, fig, rect,
         """
 
         super().__init__()
-        if isinstance(rect, mtransforms.Bbox):
-            self._position = rect
+        if "rect" in kwargs:
+            if args:
+                raise TypeError(
+                    "'rect' cannot be used together with positional arguments")
+            rect = kwargs.pop("rect")
+            _api.check_isinstance((mtransforms.Bbox, Iterable), rect=rect)
+            args = (rect,)
+        self._subplotspec = subplotspec = None
```

Comment:
```suggestion
        subplotspec = None
```
and move `self._subplotspec = None` down right before `set_subplotspec`. `self._subplotspec` is not yet needed and it's confusing to have two obviously related variables around.

I think that makes it clearer that `subplotspec` is a temporary variable that gets modified down the line and `self._subplotspec` will be updated from that through `set_subplotspec`.

Reply:
Sure.  This does by the way raise the following point: right now (with this PR) it is possible to call `set_subplotspec(spec)` on an absolute-positioned axes to convert it into a gridspec-positioned one, but `set_subplotspec(None)` won't work.  Perhaps we should try to make it work too, though we'd need to figure out exactly how it interacts with `set_position`.

Also, perhaps better terminology for "gridspec-positioned" vs "absolute-positioned" could be useful for the docs.